syne_tune/optimizer/schedulers/transfer_learning/rush.py

```diff
@@ -0,0 +1,151 @@
+
+        Reference: A resource-efficient method for repeated HPO and NAS.
+        Giovanni Zappella, David Salinas, Cédric Archambeau. AutoML workshop @ ICML 2021.
+
+        :param config_space: Configuration space for trial evaluation function.
+        :param metric: objective name to optimize, must be present in transfer learning evaluations.
+        :param transfer_learning_evaluations: dictionary from task name to offline evaluations.
+        """
+        super().__init__(config_space=config_space,
+                         transfer_learning_evaluations=transfer_learning_evaluations,
+                         metric_names=[metric])
+        points_to_evaluate = RUSHScheduler._determine_baseline_configurations(
+            config_space=config_space,
+            transfer_learning_evaluations=transfer_learning_evaluations,
+            metric=metric,
+            mode=kwargs.get('mode', 'min')
+        )
+        if 'points_to_evaluate' in kwargs:
+            points_to_evaluate += kwargs['points_to_evaluate']
+            points_to_evaluate = [dict(s) for s in set(frozenset(p.items()) for p in points_to_evaluate)]
+        kwargs['points_to_evaluate'] = points_to_evaluate
+
+        self._hyperband_scheduler = HyperbandScheduler(config_space, metric=metric, **kwargs)
+        self._num_init_configs = len(points_to_evaluate)
+        self._thresholds = dict()  # thresholds at different resource levels that must be met
+
+    @staticmethod
+    def _determine_baseline_configurations(config_space: Dict,
+                                           transfer_learning_evaluations: Dict[str, TransferLearningTaskEvaluations],
+                                           metric: str,
+                                           mode: str) -> List[Dict]:
+        argbest, best = (np.argmin, np.min) if mode == 'min' else (np.argmax, np.max)
+        baseline_configurations = list()
+        for evals in transfer_learning_evaluations.values():
+            best_hpc_idx = argbest(best(evals.objective_values(objective_name=metric)[..., -1], axis=1))
+            hpc = evals.hyperparameters.iloc[best_hpc_idx]
+            baseline_configurations.append({
+                key: hpc[key] for key in config_space
+            })
+        return baseline_configurations
+
+    def on_trial_error(self, trial: Trial) -> None:
+        self._hyperband_scheduler.on_trial_error(trial)
+
+    def on_trial_result(self, trial: Trial, result: Dict) -> str:
+        trial_decision = self._hyperband_scheduler.on_trial_result(trial, result)
+        return self._on_trial_result(trial_decision, trial, result)
+
+    def on_trial_remove(self, trial: Trial) -> None:
+        self._hyperband_scheduler.on_trial_remove(trial)
+
+    def on_trial_complete(self, trial: Trial, result: Dict) -> None:
+        self._hyperband_scheduler.on_trial_complete(trial, result)
+
+    def suggest(self, trial_id: int) -> Optional[TrialSuggestion]:
```

Comment:
Is RUSH restricted to type == 'stopping'? If not, then for promotion you probably also want to influence the resume decisions here. I mean: suggest may decide to resume a paused trial, but then RUSH may want to say: no, stay paused.

Reply:
This is getting quite complex then. Is there any chance you could implement RUSH by just implementing a new RungSystem? We currently have StoppingRungSystem, PromotionRungSystem, PASHARungSystem, CostPromotionRungSystem, so you could just do another one, no?

I could help, and if you like to refactor something there, please feel totally free.

Reply:
I will have a look into the RungSystem classes to get an idea what this is about.

Reply:
If we can implement a `RungSystem`, then its great. I would also be fine shipping the current version (asserting the promotion type is stop) and adding the rung system in another PR. @mseeger what do you think?

Reply:
That is also a possibility, it depends on Martin. I'd say the code would also be more clean if the RungSystem was implemented. Also, P-ASHA, the other contribution from Giovanni's team, did that.

Reply:
Ok awesome. I will then give a ship-it then and Martin can decide if he prefers to merge this PR (and open a new one after checking if the rung system can be extended) or to do it all at once with this PR.

Reply:
How do we access `transfer_learning_evaluations` in case we want to implement a new RungSystem? It would require us to add a new argument to HyperbandScheduler, right? @mseeger @geoalgo

Reply:
This is how P-ASHA was done, they used another value for the type argument, and then added some extra arguments. But you already have the info of transfer_learning_evaluations via points_to_evaluate, so the additional args can be simple (for example: how many of the points_to_evaluate come from transfer).

If you think this is not clean, subclassing HyperbandScheduler is another option. But I am in general not such a huge fan of making a subclass, if all I am doing is change a component of a class. Essentially, RUSH is just making a small change to the stop/continue rule, everything else stays the same.